rbac/internal/views.py

```diff
@@ -1458,3 +1460,92 @@ def retrieve_ungrouped_workspace(request):
+
+    DELETE /_private/api/utils/workspace/
+        ?id=<workspace_id> : delete a single workspace
+        (no id)            : delete all standard workspaces
+    """
+    query_params = request.GET
+    logger.info(f"Workspace list or removal: {request.method} {request.user.username}")
+
+    if request.method == "DELETE" and not destructive_ok("api"):
+        return HttpResponse("Destructive operations disallowed.", status=403)
+
+    # GET
+    if request.method == "GET":
+        if query_params.get("detail") == "true":
+            workspaces = Workspace.objects.filter(type=Workspace.Types.STANDARD)
+            serialized_ws = WorkspaceSerializer(workspaces, many=True).data
+            # Add tenant id into response
+            for ws_obj, ws_data in zip(workspaces, serialized_ws):
+                ws_data["tenant_id"] = ws_obj.tenant_id
+            payload = {"count": len(serialized_ws), "data": serialized_ws}
+            return JsonResponse(payload, status=200)
+
+        ws_count = Workspace.objects.filter(type=Workspace.Types.STANDARD).count()
+        return HttpResponse(
+            f"{ws_count} standard workspace(s) eligible for removal.", content_type="text/plain", status=200
+        )
+
+    # DELETE
+    # delete 1 standard workspace
+    if ws_id := query_params.get("id"):
+        try:
+            uuid.UUID(str(ws_id))
+        except ValueError:
+            return HttpResponse("Invalid workspace id format.", content_type="text/plain", status=400)
+
+        if not Workspace.objects.filter(type=Workspace.Types.STANDARD, id=ws_id).first():
+            return HttpResponse(
+                f"Standard workspace with id='{ws_id}' not found.", content_type="text/plain", status=404
+            )
+
+        if ws := Workspace.objects.filter(type=Workspace.Types.STANDARD, id=ws_id, children__isnull=True).first():
+            try:
+                with transaction.atomic():
+                    dual_write_handler = RelationApiDualWriteWorkspaceHandler(
+                        ws, ReplicationEventType.DELETE_WORKSPACE
+                    )
+                    dual_write_handler.replicate_deleted_workspace()
+                    ws.delete()
+                logger.info(f"Deleted workspace id='{ws_id}'")
+                return HttpResponse(f"Workspace with id='{ws_id}' deleted.", content_type="text/plain", status=200)
+            except Exception as e:
+                logger.exception(f"Workspace id='{ws_id}' deletion failed: {e}")
+                return HttpResponse(str(e), status=500)
+
+        return HttpResponse(
+            f"Workspace with id='{ws_id}' cannot be removed because it has child workspace.",
+            content_type="text/plain",
+            status=400,
+        )
+
+    # delete all standard workspaces
+    ws_count = Workspace.objects.filter(type=Workspace.Types.STANDARD).count()
+    try:
+        with transaction.atomic():
+            while True:
```

Comment:
do we need while true ?

can we just walk through workspaces  in loop ?

Reply:
related with https://github.com/RedHatInsights/insights-rbac/pull/1715#discussion_r2093334033

we remove only workspaces without children, so every round we remove only last workspaces from tree-view perspective ... we can calculate the rounds needed count after [this](https://github.com/RedHatInsights/insights-rbac/pull/1583) PR will be merged ... I can update it to not use the `while True` if you prefer

Reply:
with last commit I replaced the infinite loop by a "for" loop, after we merge https://github.com/RedHatInsights/insights-rbac/pull/1583/ we can use the `WORKSPACE_HIERARCHY_DEPTH_LIMIT ` cc: @astrozzc

Reply:
so after some discussions I did revert of the last commit and now the logic use the `while` loop again